internal/storage/fs/git/source.go

```diff
@@ -0,0 +1,159 @@
+package git
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"time"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/storage/memory"
+	"go.flipt.io/flipt/internal/containers"
+	"go.flipt.io/flipt/internal/gitfs"
+	"go.uber.org/zap"
+)
+
+// Source is an implementation of storage/fs.FSSource
+// This implementation is backed by a Git repository and it tracks an upstream reference.
+// When subscribing to this source, the upstream reference is tracked
+// by polling the upstream on a configurable interval.
+type Source struct {
+	logger *zap.Logger
+	repo   *git.Repository
+
+	url      string
+	ref      string
+	hash     plumbing.Hash
+	interval time.Duration
+	auth     transport.AuthMethod
+}
+
+// WithRef configures the target reference to be used when fetching
+// and building fs.FS implementations.
+// If it is a valid hash, then the fixed SHA value is used.
+// Otherwise, it is treated as a reference in the origin upstream.
+func WithRef(ref string) containers.Option[Source] {
+	return func(s *Source) {
+		if plumbing.IsHash(ref) {
+			s.hash = plumbing.NewHash(ref)
+			return
+		}
```

Comment:
IRC you mentioned the tests dont currently cover the git functionality, just the storage interface impl right? so im guessing these tests will be added in a later PR?

Reply:
The tests do cover the git functionality (see: https://github.com/flipt-io/flipt/pull/1705/files#diff-01ea61ca46e7a344d8ec588b8892faa33412b26822780fa1b8136c7d10d3e8a2)

But I did forget to test the `Store. Which I am doing now 👍